### PR TITLE
A robot trader using LSTM helping buying and selling stock in market

### PR DESCRIPTION
**Pull request by happynoom**
We design a solution, named DeepTrade, including history data representation, neural network construction and trading optimization methods, which could maximizing our profit based on passed experience.

### CODE REVIEW

**Comment by happynoom** [comment]
I signed it!

**Comment by EmbraceLife** [comment]
There seems to be an old use of `cmp` which is removed from python3, could you update it? thanks 

  ```
File "/Users/Natsume/Downloads/DeepTrade/rawdata.py", line 39, in read_sample_data
    sorted_data = sorted(raw_data, cmp=lambda x, y: cmp(x.date, y.date))
TypeError: 'cmp' is an invalid keyword argument for this function
Uncaught exception. Entering post mortem debugging
```

**Comment by happynoom** [comment]
Hi @EmbraceLife , thanks for your comment.
I have fixed the 'cmp' issue, and other incompatible problems.
Would you check again?
Thank you very much.

**Comment by EmbraceLife** [comment]
Thanks @happynoom it works now

**Comment by EmbraceLife** [comment]
How about add some docs and comments to major classes and methods? This way, it is easier for people to read and understand what the code is trying to do.

**Comment by EmbraceLife** [comment]
Reading `smarttrader.py`, `SmartTrader._create_loss(self)`, [see code](https://github.com/happynoom/models/blob/master/DeepTrade/smarttrader.py#L86)  Here is what I understand around the loss you defined. Please correct me if I am wrong: 

```
        self.prediction = tf.nn.relu6(norm_signal) / 6.
        self.avg_position = tf.reduce_mean(self.prediction)
        # self.cost = 0.0002 
        self.loss = -100. * tf.reduce_mean(tf.multiply((self.y - self.cost), self.prediction, name="profit"))
```
1. `self.y` refers to percentage change of prices, 
2. `self.prediction`: is to predict percentage change of prices of next day
3. `self.avg_position`: based on code, is set to be average of prediction or average of price change. But the name is defined as `avg_position`, so you define `self.avg_position` to be the percentage of positions to hold for tomorrow. right? 

In [Network Construction](https://github.com/happynoom/models/tree/master/DeepTrade),
according to your formula and explanation `Loss = -100. * mean(P * (R-c))`, Please correct me if I am wrong: 
1. P is `percentage` of position hold for tomorrow, which refers to `self.prediction` above, right?
2. R is the actual change of price of tomorrow, right?
3. c is the cost of stock transaction which is constant. 

Now, what does loss mean here? 
1. normally, we see the smaller the loss, the better trained is the model
2. loss, is usually positive and moving towards zero
3. but here when prediction is close to reality, the loss is negative. 

I could not make sense of it. Could you explain it in more detail? Thanks

**Comment by happynoom** [comment]
What you said is all right. Loss I defined in the code is actual trade loss of money. It goes negtive because the network minimize the loss of money in trade, and it start gain profit. And that is what we are willing to see. For example, I have 1 dollar, the network output 1.0, so I will buy stock with the whole dollar, next day, the stock price increase 2 percent, then I gain 1 dollar *0.02 profit, it also means I loss (-1 dollar * 0.02). If the price has no change I got nothing and loss nothing. Furthermore, I suppose trade in market has cost, such as tax. Meanwhile，if the money is used to buy like national debt, we will get constant profit. So, if I use 1 dollar buy stock, the tax and missing profit is reasonable loss, and should be added into the loss function.

**Comment by happynoom** [comment]
I think I may emphasize that the outputs of the network don't aim at the target value. They are aim at optimal position we should keep to minimize loss of money and maximize profit in trade.

**Comment by happynoom** [comment]
self.prediction: is to suggest percentage of position we should hold to next day

**Comment by EmbraceLife** [comment]
Thanks @happynoom It helps a lot! please let me rephrase myself, to make sure I understand you fine: 
0. trader.prediction is always between 0 and 1 (due to relu6), try to predict best position for tomorrow to avoid loss and maximize profit
1. trader.loss refers to actual loss of money. It can be positive and negative, being negative loss is equivalent to making profit; being positive loss is to lose money. 
2. if current position is 0.5, and prediction (for tomorrow) is 1.0, then model wants us to buy up our positions to 100% (full position) today and hold it for tomorrow; if prediction (for tomorrow) is 0.1, then model wants us to sell our position down to 10% today and keep it for tomorrow
3. given `labels` or `trader.y` are actual change prices of next day, I have checked the max and min value of `labels` (toy_stock dataset), which being 1.05 and -0.16. Does it mean there has been a day whose price rose 105% and a day whose price drop 16% compared with previous day price? or do I miss something about `labels`? 
please correct me if I am wrong, thanks.

**Comment by happynoom** [comment]
Yes， you get it quite right. The toy stock dataset is history of Shanghai composite index. It goes crazy sometimes. I want get some data in NASDAQ market, but not so easy. Actually my train data is all from market in China.

**Comment by EmbraceLife** [comment]
Thanks @happynoom , as far as I know, your toy stock data is daily price, and if it is Shanghai composite index, it should not have risen 105% or dropped 16% the next given its up or down 10% limitation, right? So, how come `trader.y` max (105%) and min (-16%)? 

By the way, you should be able to download US stock data for free from Yahoo Finance.

**Comment by happynoom** [comment]
@EmbraceLife ，up and down limitation starts at December 1996. On May 21, 1992, it rose from 616.99 to 1266.49. And on May 23, 1995, it drop from 879.42 to 750.3, crazy ha!

**Comment by happynoom** [comment]
@EmbraceLife, Save and load model features will be added in future. Or I may have to tell, what I submitted here is a simplified version, simplified network and dataset, just willing to demo my idea, and also get valuable suggestions from you experts. More docs and results and maybe source code will be published in future. Visit my website deeplearning.xin, which has Shanghai composite index advice result updated every day and profit or loss history. Thank you for all the suggestions.

**Comment by EmbraceLife** [comment]
I didn't know that! it is interesting! 
If you want valuable suggestions from experts, I think you should open source your standard project and even write a blog to increase the publicity so that more people can know it and contribute to it. This way, you are more likely to get in touch with people with similar research interest and expertise.

Personally, I think it is better for your standard project on a separate repo, rather than on tf models repo.

**Comment by happynoom** [comment]
Thank for this advice, too. I see lots of cool ideas and models on tf models repo, so the idea come to me "could I contribute to tf models?" So I public the important part in my project. It is not a product, but still works with training and predict. I am not a fund manager, I am not sure how could I deal with my project, so I public not the entire project. However, I am still willing to add the save and load features. I am on a vocation now, so it will not be these two days. Thanks for your reviews.

**Comment by talvasconcelos** [comment]
@happynoom i'm trying to train an lstm network to predict if price of a stock will rise like x% in let's say 8 hours. for training i need the output, how did you train your bot, so that he knows what to look for?

**Comment by happynoom** [comment]
@talvasconcelos see my research on my project :https://github.com/happynoom/DeepTrade and https://github.com/happynoom/DeepTrade_keras

**Comment by ConsultingSecurty** [comment]
Hello, I am Emanuel Baltensperger from Switzerland. 
I am searching for AI based on the technology of Tensorflow to use for trading on the stockmarket. 
Is your product working and is it possible to use it as a customer?

**Comment by happynoom** [comment]
@ConsultingSecurty 
see my research on my project :https://github.com/happynoom/DeepTrade and https://github.com/happynoom/DeepTrade_keras
I am concentrated on the trade model. You are welcome to join my project.

**Comment by 7Light** [comment]
Is it work ? How about BitCoin/USDT

**Comment by ConsultingSecurty** [comment]
Hello, 

thank You for inviting me to this Project. I am not a Programmer but i have a Project www.Aimetatrader.com and need to add a java for the Chart-picture to include this to a articulate intelligence. So all this has to be programmed and i only made the Frame of the Website. It is a freesource and opensource.

**Comment by tfboyd** [comment]
Closing very old PRs.  I do not like seeing work lost.  I and others are working to accept any changes to /official; but changes to other directories require the original owner to take action.  I am sorry this change died/stalled.

I am copy and pasting this message in each PR I a closing (yes kind of like an automated response); and I want to say I feel this way about each PR and wish they had ended with a merge or some other more positive finality.  

Thank you for your contribution and time.

**Comment by gordon518** [comment]
Thank  you for your post. I have downloaded all your codes and data, it can be run and the last output is "Average Train Loss at step 5000: -0.3698519 Average position 0.4867176, Validation Loss: -0.0478029 Average Position: 0.5687155". 

Does it mean that the final loss is -0.3698519 after training? In other words, is the profit 0.3698519 after training? If yes, the result isn't good, it even can't beat a simple buy and hold strategy. The profit will be 25 with the simple buy and hold strategy becuase the price started 127.61 on 1991-01-02 and ended 3195.852 on 2017-03-30. Is there any misunderstanding? Thanks.
